content/courses/connecting-to-offchain-data/oracles.md

```diff
@@ -1161,241 +1175,236 @@ Note that there can only be one escrow per user, so the above order matters.
+    const userBalanceBefore = await provider.connection.getBalance(
+      payer.publicKey,
+    );
 
-    // Send transaction
     try {
-      const tx = await program.methods
-        .deposit(amountToLockUp, failUnlockPrice)
-        .accounts({
+      const transaction = await program.methods
+        .withdraw()
+        .accountsPartial({
           user: payer.publicKey,
-          escrowAccount: escrowState,
-          systemProgram: anchor.web3.SystemProgram.programId,
+          escrowAccount: escrow,
+          feedAggregator: SOL_USD_SWITCHBOARD_FEED,
+          systemProgram: SystemProgram.programId,
         })
         .signers([payer])
         .rpc();
 
-      await provider.connection.confirmTransaction(tx, "confirmed");
-      console.log("Your transaction signature", tx);
-
-      // Fetch the created account
-      const newAccount = await program.account.escrowState.fetch(escrowState);
-
-      const escrowBalance = await provider.connection.getBalance(
-        escrowState,
-        "confirmed",
+      await confirmTransaction(
+        provider.connection,
+        transaction,
+        CONFIRMATION_COMMITMENT,
       );
-      console.log("Onchain unlock price:", newAccount.unlockPrice);
-      console.log("Amount in escrow:", escrowBalance);
 
-      // Check whether the data onchain is equal to local 'data'
-      assert(failUnlockPrice == newAccount.unlockPrice);
-      assert(escrowBalance > 0);
-    } catch (e) {
-      console.log(e);
-      assert.fail(e);
+      // Verify escrow account is closed
+      try {
+        await program.account.escrow.fetch(escrow);
+        assert.fail("Escrow account should have been closed");
+      } catch (error) {
+        console.log(error.message);
+        assert(
+          error.message.includes("Account does not exist"),
+          "Unexpected error: " + error.message,
+        );
+      }
+
+      // Verify user balance increased
+      const userBalanceAfter = await provider.connection.getBalance(
+        payer.publicKey,
+      );
+      assert(
+        userBalanceAfter > userBalanceBefore,
+        "User balance should have increased",
+      );
+    } catch (error) {
+      throw new Error(`Failed to withdraw from escrow: ${error.message}`);
     }
   });
 
-  it("Attempt to withdraw while price is below UnlockPrice", async () => {
-    let didFail = false;
+  it("creates Burry Escrow Above Current Price", async () => {
+    const solPrice: Big | null = await aggregatorAccount.fetchLatestValue();
+    if (solPrice === null) {
+      throw new Error("Aggregator holds no value");
+    }
+    const unlockPrice = solPrice.plus(PRICE_OFFSET).toNumber();
+    await createAndVerifyEscrow(unlockPrice);
+  });
 
-    // derive escrow address
-    const [escrowState] = await anchor.web3.PublicKey.findProgramAddressSync(
-      [Buffer.from("MICHAEL BURRY"), payer.publicKey.toBuffer()],
+  it("fails to withdraw while price is below UnlockPrice", async () => {
```

Comment:
I don't understand how this works, precisely how we are simulating the current price of SOL to be able to test this. Can you explain it to me?

Reply:
@mikemaccana Here is the explainer - [Anchor Toml reference ](https://www.anchor-lang.com/docs/manifest#test-validator)- 
The test environment clones the Switchboard `SOL/USD` price feed from the devnet to the local test validator.
This cloned feed maintains its last known state, providing a fixed, realistic SOL price for the tests.
Then tests interact with this local copy of the feed, fetching the SOL price as if it were live data.
While the price doesn't update in real time, it allows consistent testing against a known SOL price point. In essence, we are working with a "snapshot" of the real SOL price feed, frozen at the time of cloning when we start testing with `anchor test` command.  For this to work we have added an additional config in `Anchor.toml` file -

```toml
[test]
startup_wait = 5000
shutdown_wait = 2000
upgradeable = false

[test.validator]
bind_address = "0.0.0.0"
url = "https://api.devnet.solana.com"
ledger = ".anchor/test-ledger"
rpc_port = 8899

[[test.validator.clone]] # switchboard-solana devnet programID
address = "SW1TCH7qEPTdLsDHRgPuMQjbQxKdH2aBStViMFnt64f"

[[test.validator.clone]] # switchboard-solana devnet IDL
address = "Fi8vncGpNKbq62gPo56G4toCehWNy77GgqGkTaAF5Lkk"

# This clones the SOL/USD price feed from devnet to local test validator.
[[test.validator.clone]] # switchboard-solana SOL/USD Feed
address = "GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR"
```

This PublicKey used in tests corresponds to the cloned feed in a local environment.

```typescript
const SOL_USD_SWITCHBOARD_FEED = new PublicKey(
  "GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR"
);
```

Let me know if you still have doubts.

Reply:
I understand account cloning, but I don't understand why
`it withdraws from escrow` and `it fails to withdraw while price is below UnlockPrice` both pass when  `SOL_USD_SWITCHBOARD_FEED` is not changing.

Reply:
Okay got the concerns. Although  `SOL_USD_SWITCHBOARD_FEED` is not changing we are changing the price in test as given below to simulate the escrow behaviour.

`const unlockPrice = solPrice.plus(PRICE_OFFSET).toNumber();`

Reply:
Aha! Can you please add a comment in the tests to say this?